Xet is on the Hub blog post (#2744)
huggingface/blog: xet-on-the-hub.md

>
  Want to skip the details and get straight to faster uploads and downloads with bigger files than ever before? 

  [Click here](#ready-xet-go) to read about joining the Xet waitlist (or [head over to join immediately](https://huggingface.co/join/xet)).
</div>

Over the past few weeks, Hugging Face’s [Xet Team](https://huggingface.co/xet-team) took a major step forward by [migrating the first Model and Dataset repositories off LFS and to Xet storage](https://huggingface.co/posts/jsulz/911431940353906).

This marks one of many steps to [fulfill Hugging Face’s vision for the Hub](https://huggingface.co/blog/xethub-joins-hf) by empowering AI builders to build, iterate, and collaborate more effectively on massive models and datasets. If you're interested in deeper dives on the technology itself, check out the following posts:
* [From Files to Chunks: Improving Hugging Face Storage Efficiency](https://huggingface.co/blog/from-files-to-chunks)
* [Rearchitecting Hugging Face Uploads and Downloads](https://huggingface.co/blog/rearchitecting-uploads-and-downloads)
* [From Chunks to Blocks: Accelerating Uploads and Downloads on the Hub](https://huggingface.co/blog/from-chunks-to-blocks)

> But this post isn't about the core technology. It's a a behind-the-scenes view of getting Xet on the Hub; taking you through our proof-of-concept to the first migration of repositories. 

Reviewer: ```suggestion
But this post isn't about the core technology. It's a behind-the-scenes view of getting Xet on the Hub; taking you through our proof-of-concept to the first migration of repositories. 
```

    <figcaption>Balanced view of downloaded and sent data with reduced throughput across the board.</figcaption>
</figure>

To keep a pod from getting in this state, we put a limit to the number of concurrent uploads each machine accepts before rejecting requests that, when reached, pushes requests on to other pods in the cluster. If all pods in the cluster get in this state, then our autoscaling policies kick in. Longer-term solutions might include forcing disk syncs, removing the on-disk buffering, or switching to ephemeral storage with better throughput. 

### Migration Takeaways

These two issues led to important architectural improvements, but they weren’t the only challenges the team tackled in the weeks following the migration. We also addressed other performance issues, a memory leak (which we fixed but still can't root cause), and on advise from AWS to migrate the LFS Bridge from an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB) to a [Network Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) (NLB) to improve scalability during bursts of traffic.

> Looking back, these issues underscore a set of critical lessons: 

Reviewer 2: ```suggestion
### Lessons learned
```

> * No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  

Reviewer 2: ```suggestion
* **There is nothing like PROD**: No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  
```

> * By staging migrations incrementally and uncovering these issues before more traffic was on the infrastructure, we avoided downtime and disruption. Managing a fraction of traffic and storage made it relatively easy in comparison to if all the Hub was on Xet from day one.   

Reviewer 2: ```suggestion
* **Migrations simulate real traffic**: By staging migrations incrementally and uncovering these issues before more traffic was on the infrastructure, we avoided downtime and disruption. Managing a fraction of traffic and storage made it relatively easy in comparison to if all the Hub was on Xet from day one.   
```

> * The infrastructure and system design was hardened iteratively over weeks. Every future byte and network request on Xet will receive the benefits of these lessons. 

Reviewer 2: ```suggestion
* **Compound the learning**: The infrastructure and system design was hardened iteratively over weeks. Every future byte and network request on Xet will receive the benefits of these lessons. 
```

In short, real-world load was essential for exposing these challenges, and incremental migrations let us tackle them safely. That’s how we’re ensuring Xet-backed repositories scale reliably for the entire Hugging Face community with faster uploads, fewer bottlenecks, and minimal disruption all the way.

## Ready. Xet. Go.

With the initial migration complete, Xet is now on the Hugging Face Hub.

<p align="center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/migration.gif" alt="Xet speed" width=40%>
</p>

We’re finalizing an official integration with the `huggingface_hub` that will mean you get the benefits of Xet without any significant changes to your current workflow. To get going:

1. **Join the Waitlist**  

>    * [Sign up here](https://huggingface.co/join-xet) with your user or organization account (if you have admin/write permissions).  

Reviewer 3: Should this be https://huggingface.co/join/xet instead of https://huggingface.co/join-xet?

@jsulz @rajatarya @ylow @julien-c

Reviewer 3: If yes, I opened a PR with this proposal: #2747

Author: It should and thank you for the catch! PR approved and merged ❤️

   * Provides authentication and security guarantees  
3. The content addressed store (CAS)  
   * Enforces chunk-based deduplication for uploads and downloads.  
   * Includes an LFS Bridge for older, non-Xet clients, acting like a traditional LFS server and providing a single URL for downloads.  
4. Amazon S3  
   * Stores file contents ([blocks](https://huggingface.co/blog/from-chunks-to-blocks#scaling-deduplication-with-aggregation)) and file reconstruction metadata ([shards](https://huggingface.co/blog/from-chunks-to-blocks#scaling-deduplication-with-aggregation)).  
   * Provides the final persisted layer of data referenced by the CAS

Before moving to production, the system was launched into an ephemeral environment, where we built a [steel thread of functionality](https://www.rubick.com/steel-threads/) to enable uploading and downloading via the `huggingface_hub`. The video below shows our progress one month in: 

<video alt="Steel thread of Xet storage" controls>
    <source src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/steel-thread.mp4" type="video/mp4">
</video>

> After the heady highs of a quick proof-of-concept, the team settled into battling gnarly integration points (privacy, backward compatibility, fragmentation, etc) with the complex ecosystem that is the Hub. Eventually, the infrastructure moved into production for Hugging Face employees. With real usage now rolling in, we moved forward with the first large-scale migration.

Reviewer 4: ```suggestion
After the heady highs of a quick proof-of-concept, the team settled into battling gnarly integration points (privacy, backward compatibility, fragmentation, etc) with the complex ecosystem that is the Hub. Eventually, the infrastructure moved into production for Hugging Face team members. With real usage now rolling in, we moved forward with the first large-scale migration.
```

Given a block with an average size of 60MB and multiple partial requests to ranges within that block, the overhead accumulates. CAS reads 210MB for just 60MB delivered, a ratio of 3.5 bytes downloaded : 1 byte sent to the client. At scale across blocks of irregular sizes and ranges that can span several blocks, the 4 bytes downloaded : 1 byte sent ratio is matched. 

As usual, an image is worth a thousand words:

<figure class="image text-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/block-format-download-problem.png" alt="Graphical representation of block format issue">
    <figcaption>Red blocks are unnecessarily streamed content while green blocks show the data CAS is actually requesting and ultimately sends.</figcaption>
</figure>

To fix this, we updated the block format to store chunk-length metadata, enabling CAS to download only what each request actually needs. This format change required coordinated updates across:

* CAS APIs  
* Block metadata format  

> * All 65,000+ blocks in S3

Reviewer 5: "* 2^16 + 1 blocks on S3 (this is a coincidence I swear)"

Author: Love it

As the OS tries to flush the page cache to disk, the pod experiences throughput throttling from [block storage](https://aws.amazon.com/ebs/), causing increased latency on uploads. While uploads slowly process, new requests continue to come in, pushing more to the disk. This creates a vicious cycle, leading to a backlog of uploads and the imbalance that we saw during the migration:

<figure class="image text-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/pod-load-imbalance.png" alt="Pod load imbalance">
    <figcaption>Balanced view of downloaded and sent data with reduced throughput across the board.</figcaption>
</figure>

To keep a pod from getting in this state, we put a limit to the number of concurrent uploads each machine accepts before rejecting requests that, when reached, pushes requests on to other pods in the cluster. If all pods in the cluster get in this state, then our autoscaling policies kick in. Longer-term solutions might include forcing disk syncs, removing the on-disk buffering, or switching to ephemeral storage with better throughput. 

## Takeaways

> These two issues led to important architectural improvements, but they weren’t the only challenges the team tackled in the weeks following the migration:

Reviewer 5: I might omit the detailed bullet points and just go something like:
 - They weren’t the only challenges the team tackled in the weeks following the migration. We also addressed other performance issues, a memory leak (which we fixed but still can't root cause), and on advise from AWS changed from ALB to NLB to improve scalability.

Author: Updated!

* **Memory Leak** \- In the first few days after the migration, we noticed CAS memory usage increasing in a stepwise pattern, hinting at a leak which was fixed after investigation.
* **Slow LFS Bridge Downloads** \- While investigating prolonged download times (a reminder to always keep an eye on tail latencies), we uncovered particularly troublesome cases. Interestingly, fixing the memory leak also resolved these slow downloads.  
* **Switch from ALB to NLB** \- We eventually migrated the LFS Bridge from an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB) to a [Network Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) (NLB), improving scalability during bursts of traffic. The DNS cutover to the NLB was performed with zero downtime.

Looking back, these issues underscore a set of critical lessons: 

> * No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face employees testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  

Reviewer 4: ```suggestion
* No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  
```

* By staging migrations incrementally and uncovering these issues before more traffic was on the infrastructure, we avoided downtime and disruption. Managing a fraction of traffic and storage made it relatively easy in comparison to if all the Hub was on Xet from day one.   
* The infrastructure and system design was hardened iteratively over weeks. Every future byte and network request on Xet will receive the benefits of these lessons. 

In short, real-world load was essential for exposing these challenges, and incremental migrations let us tackle them safely. That’s how we’re ensuring Xet-backed repositories scale reliably for the entire Hugging Face community with faster uploads, fewer bottlenecks, and minimal disruption all the way.

### Ready. Xet. Go.

> With the initial migration complete, Xet is now on the Hugging Face Hub. We want to open up access more broadly, so you can feel the ease and speed of Xet storage.

Reviewer 5: "With the initial migration complete, Xet is now on the Hugging Face Hub!"

(Can drop "We want to open up access more broadly, so you can feel the ease and speed of Xet storage.", that gets repeated with the next paragraph)

Author: Dropped.

<p align="center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/migration.gif" alt="Xet speed" width=40%>

> </p>

Reviewer 4: love this
